substrate/frame/revive/src/tests.rs

```diff
@@ -1134,12 +1134,91 @@ fn delegate_call() {
 			builder::bare_instantiate(Code::Upload(caller_wasm))
 				.value(300_000)
 				.build_and_unwrap_contract();
-		// Only upload 'callee' code
-		assert_ok!(Contracts::upload_code(RuntimeOrigin::signed(ALICE), callee_wasm, 100_000,));
+
+		// Instantiate the 'callee'
+		let Contract { addr: callee_addr, .. } =
+			builder::bare_instantiate(Code::Upload(callee_wasm))
+				.value(100_000)
+				.build_and_unwrap_contract();
 
 		assert_ok!(builder::call(caller_addr)
 			.value(1337)
-			.data(callee_code_hash.as_ref().to_vec())
+			.data((callee_addr, 0u64, 0u64).encode())
+			.build());
+	});
+}
+
+#[test]
+fn delegate_call_with_weight_limit() {
+	let (caller_wasm, _caller_code_hash) = compile_module("delegate_call").unwrap();
+	let (callee_wasm, _callee_code_hash) = compile_module("delegate_call_lib").unwrap();
+
+	ExtBuilder::default().existential_deposit(500).build().execute_with(|| {
+		let _ = <Test as Config>::Currency::set_balance(&ALICE, 1_000_000);
+
+		// Instantiate the 'caller'
+		let Contract { addr: caller_addr, .. } =
+			builder::bare_instantiate(Code::Upload(caller_wasm))
+				.value(300_000)
+				.build_and_unwrap_contract();
+
+		// Instantiate the 'callee'
+		let Contract { addr: callee_addr, .. } =
+			builder::bare_instantiate(Code::Upload(callee_wasm))
+				.value(100_000)
+				.build_and_unwrap_contract();
+
+		// fails, not enough weight
+		assert_err!(
+			builder::bare_call(caller_addr)
+				.value(1337)
+				.data((callee_addr, 100u64, 100u64).encode())
+				.build()
+				.result,
+			Error::<Test>::ContractTrapped,
+		);
+
+		assert_ok!(builder::call(caller_addr)
+			.value(1337)
+			.data((callee_addr, 500_000_000u64, 100_000u64).encode())
+			.build());
+	});
+}
+
+#[test]
+fn delegate_call_with_deposit_limit() {
+	let (caller_pvm, _caller_code_hash) = compile_module("delegate_call_deposit_limit").unwrap();
+	let (callee_pvm, _callee_code_hash) = compile_module("delegate_call_lib").unwrap();
+
+	ExtBuilder::default().existential_deposit(500).build().execute_with(|| {
+		let _ = <Test as Config>::Currency::set_balance(&ALICE, 1_000_000);
+
+		// Instantiate the 'caller'
+		let Contract { addr: caller_addr, .. } =
+			builder::bare_instantiate(Code::Upload(caller_pvm))
+				.value(300_000)
+				.build_and_unwrap_contract();
+
+		// Instantiate the 'callee'
+		let Contract { addr: callee_addr, .. } =
+			builder::bare_instantiate(Code::Upload(callee_pvm))
+				.value(100_000)
+				.build_and_unwrap_contract();
+
+		// Delegate call will write 1 storage and deposit of 2 (1 item) + 32 (bytes) is required.
+		// Fails, not enough deposit
+		assert_err!(
+			builder::bare_call(caller_addr)
+				.value(1337)
+				.data((callee_addr, 33u64).encode())
+				.build()
+				.result,
+			Error::<Test>::StorageDepositLimitExhausted,
```

Comment:
I would expect `ContractTrapped` here: The `api::delegate_call` will return an error because the sub call uses too much storage. The caller then calls `unwrap` which should result in `ContractTrapped`.

I might be missing something. But it would be nice if you could get to the bottom of this.

Reply:
https://github.com/paritytech/polkadot-sdk/blob/c1238b64431b4c74d88c082099caed36177b5b11/substrate/frame/revive/src/tests.rs#L3364
Regarding to this test, a sub-call will return `StorageDepositLimitExhausted` instead of `ContractTrapped`

Reply:
In your quoted example we fail in the caller. So this is expected. In your test you are supposed to fail in the callee.

Reply:
Hmm I think we trap the caller in case the storage deposit is exhausted in the sub call. I think this is an oversight but it explains your test. I might need to revisit this.